### PR TITLE
Remove numpy dependency from the core library

### PR DESCRIPTION
**Pull request by maximsch2**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Remove numpy dependency that seemed to have been added mostly for one test to make it cleaner. Instead moved the function that needs to support both Tensor and ndarray to test utils and removed the dependency from the core library.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
`numpy` is installed with PyTorch anyway and `numpy` is everywhere so what is the motivation?

File: tests/retrieval/helpers.py [hunk 1]
```diff
@@ -30,10 +30,40 @@
...
 from tests.retrieval.inputs import _input_retrieval_scores_no_target as _irs_no_tgt
 from tests.retrieval.inputs import _input_retrieval_scores_wrong_targets as _irs_bad_tgt
-from torchmetrics.utilities.data import get_group_indexes
 
 seed_all(42)
 
+# a version of get_group_indexes that depends on NumPy is here to avoid this dependency for the full library
+
+
+def get_group_indexes(indexes: Union[Tensor, np.ndarray]) -> List[Union[Tensor, np.ndarray]]:
```
Comment (Borda) [inline tests/retrieval/helpers.py:39]: Here you still use np

**Comment by SkafteNicki** [comment]
agree with @Borda on this one, no need to remove `numpy` as requirement as it is already installed with `torch`

File: tests/retrieval/helpers.py (hunk 1 above)
Comment (maximsch2) [reply tests/retrieval/helpers.py:39]: This is in tests, we do use scikit-learn in tests as well, which is fine, I'm only worried about the core library being lean.

**Comment by maximsch2** [comment]
I don't think pytorch requires numpy to run, it's likely for tests as well. In fact, we are running pytorch in builds without numpy so it's for sure not required.

In this particular case, the core library doesn't really require numpy, so I'm thinking we should keep it lean especially given it's a no-cost thing for us. The two current uses for numpy in torchmetrics were:
- np.nan (replaced by float('nan') which is same
- helper function that was made to take np.ndarray so that the test will be used.

Both are not needed for metrics to function.

**Comment by ananthsub** [comment]
@SkafteNicki @Borda one of the concerns is around backwards compatibility for metrics. numpy introduces a large surface area which can make this harder to manage. given how minimal the usage is in torchmetrics now, we'd prefer to keep the library lean and use torch primitives instead to better control for this

**Comment by Borda** [comment]
Ok, I am fine with dropping it 🐰

**Comment by SkafteNicki** [comment]
@maximsch2 and @ananthsub great points, also fine with dropping it, did not really have a strong opinion on this :]

**Review by Borda — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Pull request merged**